### PR TITLE
[JENKINS-51603] - Relax workDirectory requirements for a JNLPLauncher constructor

### PR DESCRIPTION
**Pull request by oleg-nenashev**
See [JENKINS-51603](https://issues.jenkins-ci.org/browse/JENKINS-51603). The current API causes incompatibility between agent configurations 2.60.x and 2.73.1+ LTS baselines when you use Configuration-as-Code plugin. Although we are going to have a workaround fix there (https://github.com/jenkinsci/configuration-as-code-plugin/issues/233), it would be great to relax the API a bit

### Proposed changelog entries

* Bug: Relax API requirements in Java Web Start agent Launcher so that Configuration-as-Code plugin can load old configuration format on Jenkins 2.68 and above
  * https://github.com/jenkinsci/configuration-as-code-plugin

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@ndeloof @MadsNielsen @ewelinawilkosz @jenkinsci/code-reviewers

### CODE REVIEW

**Review by ndeloof — APPROVED** [review]

File: core/src/main/java/hudson/slaves/JNLPLauncher.java [hunk 1]
```diff
@@ -68,12 +70,22 @@
...
+     * @param tunnel Tunnel settings
+     * @param vmargs JVM arguments
+     * @param workDirSettings Settings for Work Directory management in Remoting.
+     *                        If {@code null}, {@link RemotingWorkDirSettings#getEnabledDefaults()}
+     *                        will be used to enable work directories by default in new agents.
+     * @since 2.68
+     */
     @DataBoundConstructor
-    public JNLPLauncher(@CheckForNull String tunnel, @CheckForNull String vmargs, @Nonnull RemotingWorkDirSettings workDirSettings) {
+    public JNLPLauncher(@CheckForNull String tunnel, @CheckForNull String vmargs, @Nullable RemotingWorkDirSettings workDirSettings) {
```
Comment (casz) [inline core/src/main/java/hudson/slaves/JNLPLauncher.java]: Why the mix of `Nullable` and `CheckForNull`?

**Review by jetersen — CHANGES_REQUESTED** [review]
LGTM apart from the mix of annotations

File: core/src/main/java/hudson/slaves/JNLPLauncher.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/slaves/JNLPLauncher.java]: No specific reason. I just wanted to indicate that `null` is not something we would really want to see. Can change it

**Comment by oleg-nenashev** [comment]
@casz done

**Review by jglick — COMMENTED** [review]
This is the wrong fix. `workDirSettings` should rather be made into a `@DataBoundSetter`.

**Comment by ndeloof** [comment]
@jglick I tend to agree, but doing so `workDirSettings` can't be marked final, which some developers including @oleg-nenashev seem to consider a must-have property

**Comment by jglick** [comment]
Why @oleg-nenashev? You can always make it set-once:

```java
private RemotingWorkDirSettings workDirSettings;
public RemotingWorkDirSettings getWorkDirSettings() {
    return workDirSettings != null ? workDirSettings : RemotingWorkDirSettings.getEnabledDefaults();
}
@DataBoundSetter public void setWorkDirSettings(RemotingWorkDirSettings workDirSettings) {
    if (this.workDirSettings != null) {
        throw new IllegalStateException();
    }
    this.workDirSettings = workDirSettings;
}
```

**Comment by jglick** [comment]
Fine, I guess I should just file the PR myself…

**Pull request merged**